idseq_dag/engine/pipeline_step.py

```diff
@@ -262,3 +262,14 @@ def step_resources(self):
         By default, show link to idseq-dag documentation.
         '''
         return {"IDseq Docs": "https://github.com/chanzuckerberg/idseq-dag/wiki"}
+
+    def add_optional_output(self, output_filename):
```

Comment:
Wasnt `additional_files_to_upload` and `optional_files_to_upload` (although the name is not ideal) already keep track of this files?
Probably, for this use case I would use the `additional_files_to_upload` as files that were supposed to be generated and should generate an error on the client.

Reply:
`optional_files_to_upload` is a bit strange and I think we should get rid of it. In fact when I went with a slightly different approach I removed it. The only material difference between that and `additional_files_to_upload` is that it performs a check before uploading. We could replace it with a method that adds to `additional_files_to_upload` after doing the check and it would be simplified. There is also only one usage of it so we could just remove it entirely.

I think of the optional output as a marker on top of `additional_files_to_upload` and not a replacement to it. We still need to add those files to `additional_files_to_upload` but this gives us a flag to communicate that we expect to generate a certain file. Also `additional_files_to_upload` expects full filepaths rather than just names which would be confusing to use. We could add just the names to the status, and there are some advantages to that but I think this is a bit more explicit.

Reply:
Yes, the concepts are bit hard to distinguish and have names and purposes that are hard to understand. I just fear that one extra array that takes filenames would just increase the confusion.
If you want to keep this approach, i would:
* add comments to the existing files to specify exactly how they are use (a bit as you did in comments above)
* rename optional_output ... apparently they are not optional since adding them means they are expected to be generated, right? Maybe `extra_outputs`?
* define this new array on the initialization (where you would also add the comments) for transparency